lib/widgets/new_dm_sheet.dart

```diff
@@ -0,0 +1,385 @@
+import 'package:flutter/material.dart';
+import '../api/model/model.dart';
+import '../generated/l10n/zulip_localizations.dart';
+import '../model/narrow.dart';
+import '../model/store.dart';
+import 'color.dart';
+import 'content.dart';
+import 'icons.dart';
+import 'message_list.dart';
+import 'store.dart';
+import 'text.dart';
+import 'theme.dart';
+
+void showNewDmSheet(BuildContext context) {
+  final store = PerAccountStoreWidget.of(context);
+  showModalBottomSheet<void>(
+    context: context,
+    clipBehavior: Clip.antiAlias,
+    useSafeArea: true,
+    isScrollControlled: true,
+    builder: (BuildContext context) => Padding(
+      // By default, when software keyboard is opened, the ListView
+      // expands behind the software keyboard — resulting in some
+      // list entries being covered by the keyboard. Add explicit
+      // bottom padding the size of the keyboard, which fixes this.
+      padding: EdgeInsets.only(bottom: MediaQuery.viewInsetsOf(context).bottom),
+      child: SafeArea(
+        child: PerAccountStoreWidget(
+          accountId: store.accountId,
+          child: NewDmPicker(pageContext: context)))));
+}
+
+@visibleForTesting
+class NewDmPicker extends StatefulWidget {
+  const NewDmPicker({
+    super.key,
+    required this.pageContext,
+  });
+
+  final BuildContext pageContext;
+
+  @override
+  State<NewDmPicker> createState() => _NewDmPickerState();
+}
+
+class _NewDmPickerState extends State<NewDmPicker> {
+  late TextEditingController searchController;
+  final Set<int> selectedUserIds = {};
+  List<User> filteredUsers = [];
+
+  @override
+  void initState() {
+    super.initState();
+    searchController = TextEditingController()..addListener(_handleSearchUpdate);
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _handleSearchUpdate();
+  }
+
+  @override
+  void dispose() {
+    searchController.dispose();
+    super.dispose();
+  }
+
+  void _handleSearchUpdate() {
+    final store = PerAccountStoreWidget.of(context);
+    _updateFilteredUsers(store);
+  }
+
+  // Function to sort users based on recency of DM's
+  // TODO: switch to using an `AutocompleteView` for users
+  void _updateFilteredUsers(PerAccountStore store) {
+    setState(() {
+      bool shouldIncludeSelfUser(User user) {
+        final shouldExcludeSelfUser = selectedUserIds.isNotEmpty
+          && !selectedUserIds.contains(store.selfUserId);
+
+        if (user.userId != store.selfUserId) return true;
+        return !shouldExcludeSelfUser;
+      }
+
+      filteredUsers = store.allUsers
+        .where((user) =>
+          shouldIncludeSelfUser(user) &&
+          !user.isBot &&
```

Comment:
I think we shouldn't exclude bot users. Some bots might actually utilize DMs as a way of receiving commands from the user.